cpp/include/cudf/detail/gather.cuh

```diff
@@ -678,9 +679,24 @@ std::unique_ptr<table> gather(table_view const& source_table,
                           return col.nullable();
                         });
   if (nullable) {
-    auto const op = bounds_policy == out_of_bounds_policy::NULLIFY ? gather_bitmask_op::NULLIFY
-                                                                   : gather_bitmask_op::DONT_CHECK;
-    gather_bitmask(source_table, gather_map_begin, destination_columns, op, stream, mr);
+    auto const has_nulls =
+      bounds_policy == out_of_bounds_policy::NULLIFY ||
+      std::any_of(
+        source_table.begin(), source_table.end(), [](auto const& col) { return col.has_nulls(); });
+    if (has_nulls) {
+      auto const op = bounds_policy == out_of_bounds_policy::NULLIFY
+                        ? gather_bitmask_op::NULLIFY
+                        : gather_bitmask_op::DONT_CHECK;
+      gather_bitmask(source_table, gather_map_begin, destination_columns, op, stream, mr);
+    } else {
+      for (size_type i = 0; i < source_table.num_columns(); ++i) {
+        if (source_table.column(i).nullable() and not source_table.column(i).has_nulls()) {
```

Comment:
I think in this `else` branch, you already know that `col.has_nulls()` is false for all columns, by the `std::any_of(...)`. Therefore, you should only have to check `col.nullable()` to know if you need an all-valid mask.
```suggestion
        if (source_table.column(i).nullable()) {
```

Reply:
Ah right, good point. I will update this once we have finished this CI run's cpp tests